safety_shield/scripts/urdf_capsule_generation.py

```diff
@@ -0,0 +1,527 @@
+                    print(f"Error processing fixed joint '{parent} -> {child}': {str(e)}")
+
+    plt.show()
+    return point_clouds
+
+
+def generate_joint_point_clouds(fixed_joints_and_mesh_filenames, stl_folder):
+    """
+    generate points clouds for normal robots (no fixed joints in chain)
+    :param fixed_joints_and_mesh_filenames: parent and child of each joint. Only child mesh is used
+    :param stl_folder: folder name that contains the stl meshes
+    :return: list of point clouds. One numpy array for each joint
+    """
+    point_clouds = []
+
+    # Load meshes
+    child_meshes = {}
+
+    for _, _, _, child, child_filename, child_scale in fixed_joints_and_mesh_filenames:
+        if child_filename is not None and child_filename not in child_meshes:
+            child_mesh_file = os.path.join(stl_folder, child_filename)
+            if os.path.exists(child_mesh_file):
+                child_meshes[child_filename] = trimesh.load(child_mesh_file, force='mesh')
+                child_meshes[child_filename].apply_scale(child_scale)
+
+    # Generate point clouds
+    for _, _, _, child, child_filename, _ in fixed_joints_and_mesh_filenames:
+        if child_filename is None:
+            warnings.warn("No mesh file for link" + child)
+            continue
+
+        child_mesh = child_meshes.get(child_filename)
+
+        if child_mesh is None:
+            warnings.warn("Mesh could not be loaded for: " + child)
+            continue
+
+        if child_mesh.vertices.shape[0] > 0:
+            child_points = child_mesh.vertices
+
+            point_clouds.append(child_points)
+
+    return point_clouds
+
+
+def capsule_from_points(points: np.ndarray):
+    """
+    Computes a capsule that encloses a set of points in 3D space.
+    :param points: points: An (N, 3) numpy array representing N points in 3D space.
+    :return:  A tuple containing the two endpoints of the capsule and the radius.
+    """
+    # Compute PCA of the points
+    pca = PCA(n_components=3)
+    pca.fit(points)
+    coeff = pca.components_.T
+    center = (np.amax(points, axis=0) + np.amin(points, axis=0)) / 2
+    direction = -1 * coeff[:, 0]
+    center_repmat = np.tile(center, (len(points), 1))
+    distances = np.sqrt(np.sum(np.square(
+        (center_repmat - points) - np.multiply(np.tile(np.dot((center_repmat - points), direction), (3, 1)).T,
+                                               np.tile(direction, (len(points), 1)))), axis=1))
+    distance_max = np.max(distances)
+
+    # Compute max length of the capsule
+    Z = pca.transform(points)
+    max_length = np.amax(Z, axis=0) - np.amin(Z, axis=0)
+
+    # Initialize variables
+    p1_min = center
+    p2_min = center
+    r_min = np.max(np.sqrt(np.sum((np.square(center_repmat - points)), axis=1)))
+    volume_min = 4 / 3 * math.pi * (r_min ** 3)
+
+    # Compute optimal capsule by iterating over possible endpoints
+    for i in np.arange(0.005, max_length[0] / 2, 0.0005):
```

Comment:
Where are these numbers (`0.005`, `0.0005`) coming from?
Best practice:
Define the numbers in a config file or pass them as command line arguments.
Okay practice: 
Define constant variables in beginning of the file, e.g.
`GRAVITY = 9.81`

Reply:
This code was written by Niket, and I never had to touch it. I also don't know where these values come from